### PR TITLE
[SYCL-MLIR] Merge from intel/llvm sycl branch

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Please only review https://github.com/intel/llvm/pull/10668/commits/d3ec7769aeb3fb4b7a12eaf89d019313e0750122, https://github.com/intel/llvm/pull/10668/commits/6259928c6e29486b058b00378466f81208843640, and files with conflict:
```
        both modified:   Matrix/element_wise_all_ops_impl.hpp
        both modified:   Matrix/element_wise_all_ops_tf32_impl.hpp
```

Please do not squash and merge this PR.

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
`Regression/mad_sat.cpp` fails because 
```
BuiltinBitCastExpr 0x7f65fb6ad128 'long const * __attribute__((ext_vector_type(16)))':'long const * __attribute__((ext_vector_type(16)))' <LValueToRValueBitCast>
`-DeclRefExpr 0x7f65fb6ad108 'const long *const':'const long *const' lvalue ParmVar 0x7f65fb6ace38 'from' 'const long *const &'
LValue to RValue bitcast not handled yet
```

File: sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
Comment (jopperm) [inline sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp:1]: Looks like a complete overhaul of the test -- what were our previous changes here that caused the conflict?

**Review by sommerlukas — COMMENTED** [review]
Files with conflicts and https://github.com/intel/llvm/commit/ab18fbb02049040129d7701ace67b37e21852b39 look good to me. 

The change from `vector<...>` to `std::array` in the codegen in https://github.com/intel/llvm/commit/cb9310f0b0db4da28a7c40734fe49eaa149d302b might interfere with raising, so I would like to hear @victor-eds's and @jopperm's opinion on that before merging.

File: polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp [hunk 1]
```diff
@@ -30,14 +30,17 @@ class KernelName;
 // CHECK:         sycl.host.set_captured %[[lambda]][3] = %[[bar_const]] : !llvm.ptr, i64
 
 // COM: float/double vectors are lowered to native MLIR types; again, just ensure that the value we marked comes from the CGF.
```
Comment (jopperm) [inline polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp:32]:
```suggestion
// COM: vectors are passed as pointers to a struct; check that we matched the corresponding memcpy.
```

File: polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp [hunk 2]
```diff
@@ -30,14 +30,17 @@ class KernelName;
...
+// CHECK:         %[[lambda_capture_4_gep:.*]] = llvm.getelementptr inbounds %[[lambda]][0, 4] : (!llvm.ptr) -> !llvm.ptr, !llvm.struct<{{.*}}>
 // CHECK:         %[[cgf_capture_3_gep:.*]] = llvm.getelementptr inbounds %[[cgf]][0, 3] : (!llvm.ptr) -> !llvm.ptr, !llvm.struct<"{{.*}}", (ptr, ptr, ptr, ptr, ptr)>
 // CHECK:         %[[cgf_capture_3:.*]] = llvm.load %[[cgf_capture_3_gep]] {{.*}} : !llvm.ptr -> !llvm.ptr
-// CHECK:         %[[cgf_capture_3_load:.*]] = llvm.load %[[cgf_capture_3]] {{.*}} : !llvm.ptr -> vector<4xf32>
-// CHECK:         sycl.host.set_captured %[[lambda]][4] = %[[cgf_capture_3_load]] : !llvm.ptr, vector<4xf32>
+// CHECK:         "llvm.intr.memcpy"(%[[lambda_capture_4_gep]], %[[cgf_capture_3]], {{.*}}) <{isVolatile = false}> : (!llvm.ptr, !llvm.ptr, i64) -> ()
+// CHECK:         sycl.host.set_captured %[[lambda]][4] = %[[cgf_capture_3]] : !llvm.ptr, !llvm.ptr
 
 // COM: the half vector is passed as a pointer to a struct; check that we matched the corresponding memcpy
```
Comment (jopperm) [inline polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp:39]:
```suggestion
```

**Comment by jopperm** [comment]
> The change from vector<...> to std::array in the codegen in https://github.com/intel/llvm/commit/cb9310f0b0db4da28a7c40734fe49eaa149d302b might interfere with raising

From my schedule-kernel-raising-POV it shouldn't cause problems, as we already had the memcpy-matching in place for `half` vectors, and we're not doing any data type changes during raising.

**Comment by victor-eds** [comment]
> From my schedule-kernel-raising-POV it shouldn't cause problems, as we already had the memcpy-matching in place for `half` vectors, and we're not doing any data type changes during raising.

Same, I haven't implemented constant `vec` propagation yet, so it should be fine.

File: polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp:32]: This is the new diff:
```
$ git diff upstream/sycl sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
diff --git a/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp b/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
index e394059d2b3b..61d0b4fbc5aa 100644
--- a/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
+++ b/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
@@ -71,8 +71,7 @@ void verify_op_a(const T l, const T r, const float ref, OP op) {
 
            ext::intel::experimental::matrix::joint_matrix_store(
                sg, sub_mat,
-               accessMat.template get_multi_ptr<access::decorated::no>() +
-                   (sg_startx * SUB_ROWS) * NUM_COLS +
+               accessMat.get_pointer() + (sg_startx * SUB_ROWS) * NUM_COLS +
                    sg_starty / SG_SZ * SUB_COLS,
                NUM_COLS);
          }); // parallel for
@@ -110,12 +109,11 @@ void verify_op_c(const T l, const T r, const float ref, OP op) {
              wi_slice[i] = op(wi_slice[i], r);
            }
 
-           joint_matrix_store(
-               sg, sub_mat,
-               accessMat.template get_multi_ptr<access::decorated::no>() +
-                   (sg_startx * SUB_ROWS) * NUM_COLS +
-                   sg_starty / SG_SZ * SUB_COLS,
-               NUM_COLS, layout::row_major);
+           joint_matrix_store(sg, sub_mat,
+                              accessMat.get_pointer() +
+                                  (sg_startx * SUB_ROWS) * NUM_COLS +
+                                  sg_starty / SG_SZ * SUB_COLS,
+                              NUM_COLS, layout::row_major);
          }); // parallel for
```
Basically `accessMat.template get_multi_ptr<access::decorated::no>()` is changed to `accessMat.get_pointer()`.

File: sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
Comment (whitneywhtsang) [reply sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp:1]: This is the new diff:
```
$ git diff upstream/sycl sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
diff --git a/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp b/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
index e394059d2b3b..61d0b4fbc5aa 100644
--- a/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
+++ b/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp
@@ -71,8 +71,7 @@ void verify_op_a(const T l, const T r, const float ref, OP op) {
 
            ext::intel::experimental::matrix::joint_matrix_store(
                sg, sub_mat,
-               accessMat.template get_multi_ptr<access::decorated::no>() +
-                   (sg_startx * SUB_ROWS) * NUM_COLS +
+               accessMat.get_pointer() + (sg_startx * SUB_ROWS) * NUM_COLS +
                    sg_starty / SG_SZ * SUB_COLS,
                NUM_COLS);
          }); // parallel for
@@ -110,12 +109,11 @@ void verify_op_c(const T l, const T r, const float ref, OP op) {
              wi_slice[i] = op(wi_slice[i], r);
            }
 
-           joint_matrix_store(
-               sg, sub_mat,
-               accessMat.template get_multi_ptr<access::decorated::no>() +
-                   (sg_startx * SUB_ROWS) * NUM_COLS +
-                   sg_starty / SG_SZ * SUB_COLS,
-               NUM_COLS, layout::row_major);
+           joint_matrix_store(sg, sub_mat,
+                              accessMat.get_pointer() +
+                                  (sg_startx * SUB_ROWS) * NUM_COLS +
+                                  sg_starty / SG_SZ * SUB_COLS,
+                              NUM_COLS, layout::row_major);
          }); // parallel for
```
Basically `accessMat.template get_multi_ptr<access::decorated::no>()` is changed to `accessMat.get_pointer()`.

File: polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp]: done.

File: polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/sycl/raising/host_raising_captures.cpp]: done.

**Pull request merged**